### ISSUE TITLE
Autofix fails to insert necessary parentheses

### ISSUE THREAD

**Issue by nmote** [semgrep/semgrep#6236]
**Describe the bug**

When a metavariable `$X` matches a binary expression, and the fix uses `$X` as part of another binary expression, it may be necessary to also insert parentheses around it to avoid changing the meaning of the fix.

**To Reproduce**

https://semgrep.dev/s/Kv6d

**Expected behavior**

The last example should correct to `bar((x + 5) * 2)` instead of `bar(x + 5 * 2)`

**What is the priority of the bug to you?**

- [ ] P0: blocking your adoption of Semgrep or workflow
- [ ] P1: important to fix or quite annoying
- [x] P2: regular bug that should get fixed

**Use case**
N/A

**Comment by r2c-demo** [comment]
This issue is synced in Linear at https://linear.app/r2c/issue/PA-1956/autofix-fails-to-insert-necessary-parentheses. _Note_: this link is for r2c use only and is not accessible publicly.

**Comment by aryx** [comment]
Love this! Only AST-based autofix could handle those things.

**Comment by nmote** [comment]
Fixed in https://github.com/returntocorp/semgrep/commit/f591d2e059702ad97fbb92ad1f814b89a6954c52.

I waited for a bit before closing because it uses AST-based autofix, which only worked in very specific cases at that point. However, now I think it's mature enough (at least for JS and Python) that we can close this. Currently, Semgrep will still get it wrong for other languages and for some unsupported fixes, but it should work correctly in most cases and we will continue making AST-based autofix handle more fixes.
